modules/pam_keyinit/pam_keyinit.c

```diff
@@ -234,6 +230,31 @@ int pam_sm_open_session(pam_handle_t *pamh, int flags UNUSED,
 	return ret;
 }
 
+/*
+ * Dummy
+ */
+int pam_sm_authenticate(pam_handle_t *pamh UNUSED, int flags UNUSED,
+                   int argc UNUSED, const char **argv UNUSED)
+{
+  return PAM_SUCCESS;
+}
+
+/*
+ * since setcred and open_session are called in different orders, a
+ * session ring is invoked by the first of these functions called.
+ */
+int pam_sm_setcred(pam_handle_t *pamh, int flags UNUSED,
+                   int argc, const char **argv)
+{
+  return do_open_session(pamh, argc, argv, PAM_CRED_ERR);
+}
+
+int pam_sm_open_session(pam_handle_t *pamh, int flags UNUSED,
+			int argc, const char **argv)
+{
+  return do_open_session(pamh, argc, argv, PAM_SESSION_ERR);
+}
+
```

Comment:
To be on the safe side in pam_sm_close_session I'd recommend changing `if (session_counter == 0 ... `to `if (session_counter <= 0 ...`